### ISSUE TITLE
Lambda Cold Start Time

### ISSUE THREAD

**Issue by revmischa** [prisma/prisma#10724]
### Bug description

I want to use Prisma in my AWS serverless lambda functions. I'm using Serverless Stack and [NodejsFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html) for bundling (which uses esbuild).

I enable the [minify option](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html#configuring-esbuild) when bundling.

My problem: 7.5MB of generated prisma client code is included in my application. I already have a prisma lambda layer containing the engines, my issue is with the massive amount of minified JS code included in my lambda bundle. This is an issue because it makes my lambda functions start in about 2.5 seconds compared to about 600ms for my functions that don't have the prisma client bundled:

## No prisma:
![image](https://user-images.githubusercontent.com/245131/146366389-872c37ae-2029-4627-88cd-5aceddb8bc98.png)

## With prisma:
![image](https://user-images.githubusercontent.com/245131/146366620-7a3ec580-4010-4d8d-bfde-88f8e4884f17.png)

Here is what a typical lambda bundle with prisma looks like:
![image](https://user-images.githubusercontent.com/245131/146367554-839d7d13-760b-4eb6-949b-1da456953b75.png)

I'm continuing to gather data to try to work out what fraction of my cold start time is prisma. See below comments for more details on that.

### How to reproduce

Create a NodejsFunction with AWS CDK that imports a generated prisma client.

https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html

### Expected behavior

Since I'm using esbuild (and a lot of next-gen JS tooling is starting to use it) it would be helpful if the generated client used ES modules or at least ES6 import syntax. CommonJS is death for bundling.

I don't know why the prisma clients are 7.5MB of minified JS. That seems highly strange. My schema file is about 800 lines. My `dmmfString` is 6,492,708 characters long it seems.

<img width="616" alt="Screen Shot 2021-12-16 at 14 01 59" src="https://user-images.githubusercontent.com/245131/146368124-d41828c8-16e2-442b-84fb-39535cd9ec24.png">

☝🏻  CommonJS.

https://esbuild.github.io/api/#tree-shaking

> This way esbuild will only bundle the parts of your libraries that you actually use, which can sometimes be a substantial size savings. Note that esbuild's tree shaking implementation relies on the use of ECMAScript module import and export statements. It does not work with CommonJS modules. Many libraries on npm include both formats and esbuild tries to pick the format that works with tree shaking by default. 

### Prisma information

3.6.0

### Environment & setup

- OS: linux- lambda
- Database: postgres, aurora serverless
- Node.js version: 14
 

### Prisma Version

```
prisma                  : 3.6.0
@prisma/client          : 3.6.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : dc520b92b1ebb2d28dc3161f9f82e875bd35d727
Studio                  : 0.440.0
```

**Comment by janpio** [comment]
Hey @revmischa, you seem to have a pretty good setup already. Can you maybe create a Prisma Client with a minimal schema file (1 tiny model with an ID and a string field for example) and let us know what the impact on the Lambda initialization of that is? That would be super helpful to see an additional data point to "no Prisma" and "Prisma with massive `dmmfString`".

**Comment by revmischa** [comment]
Here's an empty schema in a new CDK/SST project using NodejsFunction

![image](https://user-images.githubusercontent.com/245131/146399917-d96915f2-92da-4908-a1f1-5d7bc4c41765.png)


Can find the repo here: https://github.com/jetbridge/prisma-bundle-size

**Comment by janpio** [comment]
Sorry, should have been clearer: The Lambda timeline, Initialization measurement would be interesting for this minimal schema.

**Comment by revmischa** [comment]
Okay I created two lambdas, one with prisma:
```ts
import { PrismaClient } from "@prisma/client";

export async function handler() {
  const client = new PrismaClient();
  console.log(await client.coolTable.findMany());
}
```

And another empty lambda.

For the prisma lambda I get this error:
`error: Error validating datasource `db`: the URL must start with the protocol postgresql:// or postgres://."`
Which is more or less what I expect.

The "initialization" section is 387ms, but then 2.4s for the invocation, which seems suspicious:
![image](https://user-images.githubusercontent.com/245131/146425677-c6ceb8d4-ad8a-4ef3-abfe-d6f1ba616451.png)

As a control I created an empty lambda which looked like this:
![image](https://user-images.githubusercontent.com/245131/146425840-f47a086d-42d4-42b3-a829-4283f4015576.png)

All the code is in the sample repo.

**Comment by revmischa** [comment]
Related question: I see that my app is bundling `@prisma/client/runtime/index.js`. I also see that there is an `esm` directory there with an ESM version of index.mjs. I would like to use the ESM version, but I'm unclear on how to do this.

The `package.json` contains: 
```
  "main": "index.js",
```
But it has no exports section, so I'm not sure how normal module bundling could know that the esm version exists. Should @prisma/client/runtime's package.json contain an [exports section as recommended](https://nodejs.org/api/packages.html#dual-commonjses-module-packages) in the node docs?

E.g.
```
{
  "main": "./index.js",
  "exports": {
    "import": "./esm/index.mjs",
    "require": "./index.js"
  }
}
```

**Comment by janpio** [comment]
> Related question

Please open a new issue to ask that question - we can not handle two different discussions in one issue without things getting complicated. Thanks.

**Comment by revmischa** [comment]
Trying to gather more data. I want to know how much the dmmfString adds to startup time.

I tested setting `dmmfString = "{}"` to compare bundling and profiling and see how the giant JSON string affects my function.


## Run time, startup time, trace, and bundle size with empty dmmfString

```
bash-4.2# time node index.js

real    0m1.765s
user    0m0.772s
sys     0m0.415s
```

<img width="1786" alt="emptyDmmfFire" src="https://user-images.githubusercontent.com/245131/146932428-d2d55702-bd67-466b-8ef4-351654ff7b89.png">


<img width="1171" alt="emptyDmmf" src="https://user-images.githubusercontent.com/245131/146932439-441637b4-1d51-41d3-a4c6-fc6fd055d7c5.png">


<img width="1772" alt="Screen Shot 2021-12-21 at 14 15 09" src="https://user-images.githubusercontent.com/245131/146928677-0b42466e-f5a5-4bad-be49-1f1905946cca.png">


<hr/>

## Same setup with full `dmmfString`

```
bash-4.2# time node index.js

real    0m2.175s
user    0m1.058s
sys     0m0.462s
```

<img width="1780" alt="Screen Shot 2021-12-21 at 14 52 11" src="https://user-images.githubusercontent.com/245131/146933054-31b1009e-09e2-4983-b5b8-7d8cce9c5f2b.png">



So in conclusion I would say just the parsing of the giant dmmfString alone adds about 400ms. I don't think it's the only slowness; the 2.4s invocation of the repro lambda that does nothing but import prisma and try a query is concerning. I will try to continue gathering data but if you have any suggestions of things to try I'd be happy to give them a go.

**Comment by janpio** [comment]
> the 2.4s invocation of the repro lambda that does nothing but import prisma and try a query is concerning

How much time does opening a connection to the database take? That is usually the slow part.

**Comment by revmischa** [comment]
> `error: Error validating datasource `db`: the URL must start with the protocol postgresql:// or postgres://."`

I didn't pass a valid DSN so it just errors immediately with `error: Error validating datasource db: the URL must start with the protocol postgresql:// or postgres://."`

**Comment by revmischa** [comment]
Continuing to profile prisma during nodejs startup. According to this profile, 48% of my ~1.5s startup time is initializing the prisma library, with significant time spent on "findSync"


<img width="1312" alt="Screen Shot 2021-12-27 at 2 48 16 PM" src="https://user-images.githubusercontent.com/245131/147473453-4be679eb-c821-4cda-ac58-6d5b444dafa4.png">

<img width="1302" alt="Screen Shot 2021-12-27 at 2 48 47 PM" src="https://user-images.githubusercontent.com/245131/147473457-f2b88bc4-4dca-40a3-8b2d-8be1c0067622.png">

<img width="1396" alt="Screen Shot 2021-12-27 at 2 50 22 PM" src="https://user-images.githubusercontent.com/245131/147473461-189b44e1-12e8-4a7d-9614-2b230cd554eb.png">

**Comment by Jolg42** [comment]
Sounds to me that the `findSync` here is the following (recursive) function
https://github.com/prisma/prisma/blob/main/packages/client/src/runtime/utils/find.ts#L46-L122

Which is called once when the Prisma Client is initialized
https://github.com/prisma/prisma/blob/main/packages/client/src/generation/utils/buildDirname.ts#L19-L42

From 1st post description
>I'm using Serverless Stack and NodejsFunction for bundling (which uses esbuild).
We are in an bundler case here.

**Comment by pantharshit00** [comment]
I think https://github.com/prisma/prisma/pull/10804 might help here as well.

**Comment by revmischa** [comment]
Thanks @pantharshit00, will be totally happy to profile it when it's released.

I suppose that likely leaves my giant DMMF string parsing time as at least one of my biggest problems. Any ideas about any way I can speed that up? Or anything I am doing that would cause my dmmfString to be 6,492,708 characters long? Is that normal?

**Comment by pabacham** [comment]
Let me duplicate the report here

Framework: serverless-stack (SST) '@prisma/client' and '.prisma' are not a part of bundle, Lambda takes them from a layer.

Prisma layer:
Folders and files structure - https://monosnap.com/file/qrVf8tnkUeV3PxM3Gn8ilP41x9mQNz
Total size - 65mb
Compressed size - 17mb
Prisma Schema is around 15k lines

Report for simple lambda function with conditions:

Simple Lambda function
Function - https://monosnap.com/file/0xtqAJJ7lluJIJJI5GNdwJ3pgDxr0x
Report - https://monosnap.com/file/6QNcLCn3gkEQ74yQr5WzyKeLdilZfh

Lambda function with attached Prisma Layer + imported Prisma Client in code
Function - https://monosnap.com/file/2WGONdIqGAvgm1Z1hl0K2Myo0oh84B
Report - https://monosnap.com/file/BwE5J2FHrDoUqguSBVb7I1yb00aPzv

Lambda function with attached Prisma Layer + imported and defined Prisma Client in code
Function - https://monosnap.com/file/sm15c0UBbo8JcZgHo2YQiOIcPKHWjJ
Report - https://monosnap.com/file/eKSmk06r7Ld0ceqwF0WPsxXDikYSH1

Lambda function with attached Prisma Layer + imported and defined Prisma Client in code + some SQL queries
Function - https://monosnap.com/file/TScch5bDZ9qdWr7TMKaKWlBgJocsIz
Report - https://monosnap.com/file/czpWPAud12XPOZ3kOkX4dVsdscbk1K

**Comment by pat-producepay** [comment]
> Okay I created two lambdas, one with prisma:
> 
> ```ts
> import { PrismaClient } from "@prisma/client";
> 
> export async function handler() {
>   const client = new PrismaClient();
>   console.log(await client.coolTable.findMany());
> }
> ```
> 
> And another empty lambda.
> 
> For the prisma lambda I get this error: `error: Error validating datasource `db`: the URL must start with the protocol postgresql:// or postgres://."` Which is more or less what I expect.
> 
> The "initialization" section is 387ms, but then 2.4s for the invocation, which seems suspicious: ![image](https://user-images.githubusercontent.com/245131/146425677-c6ceb8d4-ad8a-4ef3-abfe-d6f1ba616451.png)
> 
> As a control I created an empty lambda which looked like this: ![image](https://user-images.githubusercontent.com/245131/146425840-f47a086d-42d4-42b3-a829-4283f4015576.png)
> 
> All the code is in the sample repo.

I know this is a bit dated so feel free to ignore this if you've already found a solution. So I noticed in this example your "cold start" time is in your invocation not your initialization presumably because you're establishing the db connection within the handler right?

While it might not be a silver bullet, you can try instantiating your client outside the handler. It won't solve the initial cold start, but your subsequent queries will be able to reuse your existing DB connection until amazon decides to tear down the container. https://www.jeremydaly.com/reuse-database-connections-aws-lambda/

You could probably get creative some how and long-live an instantiated client that gets reuse across lambdas, but I haven't given that solution much thought.

**Comment by renchap** [comment]
We are seeing similar problems with cold start on Google Cloud Run (Google Cloud container platform with scale-to-zero feature).
After doing some quick and dirty instrumentation, it seems that `require("prisma") can take up to 8 seconds for us. Our generated Prisma file is about 10 MB, and I think the delay might be caused by Node parsing and compiling to JIT all this code.
Note: we are still using Prisma 2.

Unfortunately, I havent found a way to properly measure the `require` times with Node yet, but this might be something you want to look for.

A simple way to measure it doing something similar to this as soon as possible in your app/function startup:
```
console.time('prisma-require');
const { PrismaClient } = require('@prisma/client');
console.timeEnd('prisma-require');
```
Caution: you need to ensure your bundler/transpiler does not transform this or move the code around for it to work

I will try to get more data about this, but this may be a clue about what is happening.

**Comment by revmischa** [comment]
> I will try to get more data about this, but this may be a clue about what is happening.

How many lines is your prisma schema and how big is your generated client `dmmfString` var?

**Comment by revmischa** [comment]
> > How many lines is your prisma schema
> 
> 1600 lines
> 
> > how big is your generated client `dmmfString` var?
> 
> ```
> grep "const dmmfString" node_modules/.prisma/client/index.js | wc -c
>  9 864 027
> ```

Yeah I would assume parsing 10MB of JSON may be slowing things down. Not to mention blowing up your function bundle sizes.

**Comment by renchap** [comment]
We are running containers, with a standard express stack on Node 16 transpiled by SWC, so no issues with bundling here.
But the CPUs on Cloud Run dont have much power, so we see 5-8s delay between the NodeJS invocation (on container start) and our app being fully loaded. On our development M1 macbooks, it takes 2-3 seconds, which is still noticeable.

What would be the best way to instrument (in production) the Prisma initialisation process, to see what takes time here?

Are there any plans to switch to something else than a JSON string parsed at startup for the DMMF schema? It would be awesome to have it generated as plain JS (and/or Rust) at app build time.

**Comment by revmischa** [comment]
I've been using the '0x' tool to instrument my function

**Comment by janpio** [comment]
Can you split that information out into its own (cross-linked) issue @renchap? That would indeed be super useful for us. Thanks!

**Comment by pabacham** [comment]
> Thanks for the tips. My problem is the cold start though. I'm building a serious application with Prisma on Lambda, so it's split into dozens of functions. The cost for provisioned concurrency is $15/mo/function/concurrency so it's not really a viable solution for me.
> Some of that time is DB connection setup overhead and presumably TLS, although that should not take seconds. Some is possibly lazy loading as well, I don't know enough about Prisma internals to say.

@revmischa Instead of running PC try lambda-warmer, it works extremely cost-efficiently.
Lambda handler
```
import warmer from "lambda-warmer";

... <in handler>

if (await warmer(event)) return parseResult('warmed');
```
Cron helper using SST
```
import {Function as sstFunction, Cron} from "@serverless-stack/resources";

...

addReservedConcurrency(fn: sstFunction, fnName: string, dev = 1, prod = 1): void {
        if (this.isNamedEnv) {
            new Cron(this, `${fnName}WarmUp`,
                {
                    schedule: "cron(0/5 6-20 * * ? *)",
                    job: {
                        function: fn,
                        jobProps: {
                            event: RuleTargetInput.fromObject({
                                "warmer": true,
                                "concurrency": (this.isProd ? prod : dev)
                            })
                        }
                    }
                })
        }
    }
}
```

Rough comparison
![image](https://user-images.githubusercontent.com/6135444/156344884-abb24fa4-2dda-4631-a40c-2fe70bd953de.png)

But still, it would be great to finally get some results from the Prisma team regarding the Cold Starts issue and using AWS Proxy or at least grant access to their own one to the projects that are out of Github.

**Comment by millsp** [comment]
Hey @revmischa, if you use the `dataproxy`, we currently compress the dmmf with `lz` string compression. We have thought of enabling this for everyone but `lz-string` is actually slow for very huge dmmf strings. It could be viable if we can find a more efficient string compression library.

**Comment by revmischa** [comment]
> 

I saw that but I don't see how that would decrease startup time. That would simply add an extra decompression step to the client startup, which I assume would only make it slower? It would still have to be decompressed and then parsed. Maybe it would help with bundle size yes, but not startup time.

I'm proposing to make the actual DMMF data structure itself by not serializing `false` values. Starting with this: https://github.com/prisma/prisma/pull/12454 and hopefully some other people who know the data model better can propose other ways to remove explicit information that can be made implicit.

**Comment by millsp** [comment]
> but not startup time

Indeed, sorry for the confusion! We're also starting to explore ways to have getDmmf as a Wasm module, that could maybe remove the need to write the DMMF in the index in the first place. We might be able to do it today directly with the current getDmmf impl. at the cost of the Data Proxy not being able to benefit from it.

**Comment by n3oney** [comment]
You should try the new `4.11.0` release, with the `jsonProtocol` preview feature enabled.

I've had a similar issue (#18104), and can confirm that this helped!

**Comment by janpio** [comment]
Useful context in the release notes: https://github.com/prisma/prisma/releases/tag/4.11.0

**Comment by revmischa** [comment]
After deploying prisma 4.12 with `jsonProtocol` and some [tree-shaking improvements](https://spiegelmock.com/2023/04/02/mastering-javascript-tree-shaking/) I got my cold start time down by around 200ms or 1/5th. Some improvement but incremental improvement.

You can enable [Lambda Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights.html) on your functions and it will log the cold start times for functions. 

![image](https://user-images.githubusercontent.com/245131/230993955-6e78c1bf-6515-435a-9f62-65323ee43c4d.png)

**Comment by janpio** [comment]
We recently wrote a nice blog post about `jsonProtocol` and the other work we have been doing to improve the cold start impact of using Prisma Client: [How We Sped Up Serverless Cold Starts with Prisma by 9x](https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t)

[<img width="657" alt="image" src="https://user-images.githubusercontent.com/183673/234291600-7c8201da-80cc-4238-bf98-687500ea37f8.png">](https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t)

This also includes steps like making sure the database is hosted in the same region than your serverless function.

---

To see where we are at, I also re-tested the cold start problems with specifically `require` for Google Cloud Run that @renchap reported first here and then split out into its own issue: https://github.com/prisma/prisma/issues/12132

[![image](https://user-images.githubusercontent.com/183673/234291740-606e8c49-f5df-45ba-a15e-a38bec550545.png)
](https://github.com/prisma/prisma/issues/12132#issuecomment-1521762863)
(the blue block is the `require` timing)

---

Optimally you could give us another set of data @revmischa @pabacham on how the current version 4.13.0 changes the performance of your functions. We'll keep working on this topic and really try to optimize the latency using Prisma adds here.

**Comment by janpio** [comment]
I am closing this issue for now. 

We further improved the cold start time after our blog post and my comment here, and are now generally happy with the cold start time even with massive schemas. `jsonProtocol` will soon be promoted to GA, and with that the default experience of using Prisma should have no negative impacts on cold starts in serverless environments any more compared to other database Clients.

If someone thinks their app still has slow cold start times, please open a new issue and we will be happy to take a look and figure out what is happening. Thanks!
